Support simple assembly name in startup hook

vitek-karas opened the pull request:
Implements support for using simple assembly names in the startup hook environment variable.
The code first tests if the value is a fully qualified path, and if so, will load the assembly by path.
If it's not, it will parse the value with `new AssemblyName`. If that works, it will load the assembly by name.

Note that the code still blocks relative assembly paths, as those will not parse as assembly names.

The change is a bit larger to avoid parsing the assembly name twice.

Tests will be added in the core-setup repo, since all the existing startup hook tests are there as well.

Fixes #23343

jkotas commented on src/System.Private.CoreLib/src/System/StartupHookProvider.cs:
```diff
@@ -31,34 +34,52 @@ private static void ProcessStartupHooks()
             {
-                if (String.IsNullOrEmpty(startupHook))
+                string startupHookPart = startupHookParts[i];
+                if (string.IsNullOrEmpty(startupHookPart))
```
Nit: I am wondering whether we should skip empty parts here instead of throwing to make it easier for folks building string with multiple hooks - they will be able to always append PathSeparator.

vitek-karas replied:
I'll let @sbomer comment on that. The current code is very strict, fails on trailing path separator as well.

jkotalik replied:
To follow up, if the method `ProcessStartupHooks` throws, does that cause startup to fail?

jkotas replied:
Yes

sbomer replied:
Yes, it's currently pretty strict, but I have no objections to relaxing this to make it easier to use.

jkotas replied:
Any plans to fix this? (It is 3 line change...)

vitek-karas replied:
Sorry - added this functionality in the most recent commits.
It now allows extra path separators. It still doesn't allow whitespaces, I considered it, but it felt wrong.

sbomer commented on src/System.Private.CoreLib/src/System/StartupHookProvider.cs:
```diff
@@ -85,10 +106,10 @@ private static void CallStartupHook(string assemblyPath)
-                    // Found multiple
+                    // Found multiple. Will throw below due to initializeMethod being null.
                     Debug.Assert(initializeMethod == null);
-                    wrongSignature = true;
```
This actually looks to me like either an incorrect assert that was never hit for some reason, or some kind of bug. The intention is for this case (multiple startup hook methods with incorrect signatures) to throw an ArgumentException, tested in https://github.com/dotnet/core-setup/blob/master/src/test/HostActivationTests/StartupHooks.cs#L435 - I don't know how this test is passing currently.

vitek-karas replied:
The assert is a bit redundant, but not wrong - it should be null when we get there.
This line removal was actually suggested by the compiler - it marked the line as unnecessary. Which is true, the `wrongSignature` is not looked at if `initializeMethod` is `null`.

sbomer commented on src/System.Private.CoreLib/src/System/StartupHookProvider.cs:
```diff
@@ -31,34 +34,52 @@ private static void ProcessStartupHooks()
+                }
+                else
                 {
-                    throw new ArgumentException(SR.Argument_AbsolutePathRequired);
```
It might be nice for diagnostic purposes to continue throwing an ArgumentException for relative paths. Or, if we fall back to exceptions thrown by the AssemblyName constructor, we should update the error handling section of the spec.

vitek-karas replied:
Now it's throwing ArgumentException for anything but simple names (and wraps exceptions from AssemblyName ctor).

sbomer commented:
I think we should note some likely pitfalls and provide an example in the spec of using the assembly name. At first glance it looks like this requires the assemblyname to be `'MyAssembly'` rather than `'MyAssembly.dll'` for it to be loaded correctly - do you know if that's true?

Also, I played around with the `AssemblyName(string)` ctor, and found that it accepts inputs like:
- `MyAssembly.dll.dll, Version=1.2.3, something=nothing`
- `MyAssembly.dll.dll something`
- `MyAssembly ` (note the space).

I'm wondering whether it would make sense to perform some checks on the input before giving it to `AssemblyName`, to give better diagnostics - some of these seem more like syntax errors to me. Another option might be to improve the errors produced by `AssemblyName`.

vitek-karas commented:
@sbomer I'm kind of split on this (and yes I thought about it when I was writing the code :-)).

`Assembly.dll` - this is a tricky one. We can't rule out dots in assembly names (almost every assembly has it). So the only thing we could do is rule out names which end with `.dll` (any casing). That might actually be useful (unless you have an assembly with that name, but then it's probably your fault).
More complex names - also tricky
* We could simply rule out anything which has a comma or space in it. That would effectively restrict to simple assembly names.
* It means there's no way to specify version, public key - culture is probably not interesting as that usually means satellite assembly.. not a good one for startup hook.
* On the other hand, since the assembly must come from TPA (unless you have another startup hook providing assembly resolution), then specifying version and such is sort of counterproductive anyway.

I guess I like it in the end - it would mean that we intentionally fail on stuff which looks like more than a simple name. That also means that in the future we could decide to support more without breaking anything.

Thoughts?

sbomer commented:
I like the policy of only supporting simple names until we see a reason to support full names (I can't think of one off the top of my head). For the `Assembly.dll` case, since it is technically a valid simple name I'd prefer to support it, but make it absolutely clear in the spec that the hook will need to be the assembly name without extension.

vitek-karas commented:
I actually think that we should not support `Assembly.dll`. It's the most likely form people would try if they want relative path support. And it will NOT work, and will give seemingly wrong exception message. The exception message will definitely not lead to "ah, it does not support relative paths".
Specs are great, better yet if they make it into docs... but in reality very few people actually read either.

sdmaclea commented:
@vitek-karas @sbomer Sven's initial review commented on the removal of `IsPartiallyQualifiedPath()` maybe it needs to be part of the handling of `Assembly.dll`.  Not sure now much logic it is worth, but if the if it is a partially qualified path to a file that exists, but fails to load as an assemblyname, then it seems logical it should be an exception about a path.

sbomer commented:
@sdmaclea agreed - or if we don't want to check for file existence, at least hooks that are relative paths but not valid assembly names, like './Assembly.dll' could benefit from such an exception.

sbomer commented:
@sdmaclea we already document and test the behavior of multiple hooks - if you're referring to @jkotas's comment, I believe that was about relaxing the syntax to allow empty entries in the list.

sdmaclea commented:
@sbomer I was referring to this comment
> To follow up, if the method ProcessStartupHooks throws, does that cause startup to fail?

vitek-karas commented:
I don't like checking for file existence to provide a better error message. It basically means we would be basic the validity of the string on file existence... feels wrong.

Currently I'm inclined to simply disallow any directory separator, space and comma as well as check for the `.dll` extension. I know that it would prohibit assemblies with name `Assembly.dll`, but I think that's acceptable compromise to get a very predictable behavior.

I'm fine with wrapping the assembly loading code in a try/catch and wrap the exception in something about startup hooks.

vitek-karas commented:
I don't think it's currently possible for one startup hook to catch exceptions thrown from another one (and potentially let the program run). I also don't think this is something we should allow, at least for now.

@sdmaclea is that the scenario you had in mind?

sdmaclea commented:
>  I don't think it's currently possible for one startup hook to catch exceptions thrown from another one > @sdmaclea is that the scenario you had in mind?

@vitek-karas ASP.NET wants to add a startup hook which registers an unhandled exception handler.  The handler will put up a webpage possibly with info about the **fatal** error. The process would ultimately terminate.  I believe it should be possible.  I just want to make sure they can configure it so that it is the first handler installed.

sdmaclea commented:
I would rather support assemblies with '.' in the name.  I have been fighting with errors the last week about the facade Assembly `System.Runtime.Loader`.  I believe they will be common in Assembly names in some organizations.

vitek-karas commented:
Implemented the simple name checks (no directory separator, command or space and no `.dll`).
Added exception wrapping.

The related dotnet/core-setup#5597 has update tests for this as well.

jkotas commented on src/System.Private.CoreLib/src/System/StartupHookProvider.cs:
```diff
@@ -17,6 +14,22 @@ internal static class StartupHookProvider
         private const string StartupHookTypeName = "StartupHook";
         private const string InitializeMethodName = "Initialize";
 
+        private static readonly char[] s_disallowedSimpleAssemblyNameChars = new char[]
```
No need to cache this for the duration of the process. It is used exactly once during startup.
